File: store/socketManager.js
// src/store/socketManager.js

let metricsSocket = null;
let liveDataSocket = null;

// Track the currently ACTIVE socket ID (e.g., "5t", null)
let activeMetricsId = undefined; 
// Track the LATEST requested ID
let pendingMetricsId = undefined;

// Timer for debouncing (The Fix)
let connectTimer = null;

let metricsChain = Promise.resolve();
let liveDataChain = Promise.resolve();

const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export const socketManager = {
  // -------------------------
  // LIVE DATA HANDLERS (Persistent)
  // -------------------------
  connectLiveData: (createSocketFn) => {
    liveDataChain = liveDataChain.then(async () => {
      if (liveDataSocket) return; // Keep alive
      liveDataSocket = createSocketFn();
    }).catch(err => console.error("LiveData Chain Error:", err));
  },

  closeLiveDataSocket: () => {
    liveDataChain = liveDataChain.then(async () => {
      if (liveDataSocket) {
        liveDataSocket.close();
        liveDataSocket = null;
      }
      await wait(100);
    });
  },

  // -------------------------
  // METRICS HANDLERS (Debounced)
  // -------------------------
  
  connectMetrics: (id, createSocketFn) => {
    // 1. Update the latest desired ID
    pendingMetricsId = id;

    // 2. CANCEL any pending connection that hasn't started yet.
    // This stops the "13.5m -> 12m -> 7m" rapid-fire chain.
    if (connectTimer) {
      clearTimeout(connectTimer);
      connectTimer = null;
    }

    // 3. Start a timer. We only execute if no new requests come in for 250ms.
    connectTimer = setTimeout(() => {
      
      metricsChain = metricsChain.then(async () => {
        // Double-check: Is this still the latest ID?
        if (pendingMetricsId !== id) return;

        // 4. Close existing socket (if any)
        if (metricsSocket) {
          // console.log(`[SocketManager] Closing previous socket (${activeMetricsId}) to open ${id}`);
          metricsSocket.close();
          metricsSocket = null;
          activeMetricsId = undefined;
          await wait(300); // Wait for clean disconnect
        }

        // 5. Connect (Only if we are still the latest request)
        if (pendingMetricsId === id) {
          console.log(`[SocketManager] Connecting Metrics Socket: ${id === null ? "ALL (null)" : id}`);
          activeMetricsId = id;
          metricsSocket = createSocketFn();
        }
      }).catch(err => console.error("Metrics Chain Error:", err));

    }, 250); // 250ms debounce delay
  },

  closeMetricsSocket: (id) => {
    metricsChain = metricsChain.then(async () => {
      // Only close if we are actually connected to this ID
      if (metricsSocket && activeMetricsId === id) {
        console.log(`[SocketManager] Closing Metrics Socket for ${id}.`);
        metricsSocket.close();
        metricsSocket = null;
        activeMetricsId = undefined;
      }
      await wait(100);
    });
  },

  // -------------------------
  // GLOBAL CLEANUP (Logout)
  // -------------------------
  closeAllSockets: () => {
    console.log("[SocketManager] FORCE CLOSING ALL SOCKETS.");
    
    // Clear any pending debounce timers
    if (connectTimer) {
      clearTimeout(connectTimer);
      connectTimer = null;
    }

    liveDataChain = Promise.resolve();
    metricsChain = Promise.resolve();

    if (liveDataSocket) {
      liveDataSocket.close();
      liveDataSocket = null;
    }
    if (metricsSocket) {
      metricsSocket.close();
      metricsSocket = null;
      activeMetricsId = undefined;
      pendingMetricsId = undefined;
    }
  },
};

File: flutter_dashboard/lib/src/services/socket_manager.dart
import 'dart:async';
import 'dart:convert';

import 'robust_web_socket.dart';

class SocketManager {
  RobustWebSocket? _metricsSocket;
  RobustWebSocket? _liveDataSocket;

  String? _activeMetricsIdKey;
  String? _pendingMetricsIdKey;

  Timer? _connectTimer;

  Future<void> _metricsChain = Future<void>.value();
  Future<void> _liveDataChain = Future<void>.value();

  Future<void> _wait(int milliseconds) {
    return Future<void>.delayed(Duration(milliseconds: milliseconds));
  }

  String _toIdKey(Object? id) {
    try {
      return jsonEncode(id);
    } catch (_) {
      return id.toString();
    }
  }

  void connectLiveData(RobustWebSocket Function() createSocketFn) {
    _liveDataChain = _liveDataChain.then((_) async {
      if (_liveDataSocket != null) {
        return;
      }
      _liveDataSocket = createSocketFn();
    }).catchError((_) {
      // keep chain alive
    });
  }

  void closeLiveDataSocket() {
    _liveDataChain = _liveDataChain.then((_) async {
      _liveDataSocket?.close();
      _liveDataSocket = null;
      await _wait(100);
    }).catchError((_) {
      // keep chain alive
    });
  }

  void connectMetrics(
    Object? id,
    RobustWebSocket Function() createSocketFn,
  ) {
    _pendingMetricsIdKey = _toIdKey(id);

    _connectTimer?.cancel();
    _connectTimer = Timer(const Duration(milliseconds: 250), () {
      final requestKey = _toIdKey(id);
      _metricsChain = _metricsChain.then((_) async {
        if (_pendingMetricsIdKey != requestKey) {
          return;
        }

        if (_metricsSocket != null) {
          _metricsSocket!.close();
          _metricsSocket = null;
          _activeMetricsIdKey = null;
          await _wait(300);
        }

        if (_pendingMetricsIdKey == requestKey) {
          _activeMetricsIdKey = requestKey;
          _metricsSocket = createSocketFn();
        }
      }).catchError((_) {
        // keep chain alive
      });
    });
  }

  void closeMetricsSocket(Object? id) {
    final idKey = _toIdKey(id);
    _metricsChain = _metricsChain.then((_) async {
      if (_metricsSocket != null && _activeMetricsIdKey == idKey) {
        _metricsSocket!.close();
        _metricsSocket = null;
        _activeMetricsIdKey = null;
      }
      await _wait(100);
    }).catchError((_) {
      // keep chain alive
    });
  }

  void closeAllSockets() {
    _connectTimer?.cancel();
    _connectTimer = null;

    _liveDataChain = Future<void>.value();
    _metricsChain = Future<void>.value();

    _liveDataSocket?.close();
    _liveDataSocket = null;

    _metricsSocket?.close();
    _metricsSocket = null;
    _activeMetricsIdKey = null;
    _pendingMetricsIdKey = null;
  }
}
